packages/ui-kit/src/components/AccessTokenProvider/AccessTokenProvider.tsx

```diff
@@ -0,0 +1,131 @@
+import React, { createContext, PropsWithChildren, useCallback, useEffect, useRef, useState } from 'react'
+import { useLog } from '../Log'
+
+export interface AccessTokenContextValue {
+  /**
+   * The returned accessToken wheter from the props or from the fetchToken function
+   */
+  accessToken?: string
+  /**
+   * If true, the access token is being fetched
+   */
+  isLoading?: boolean
+  /**
+   * Function that will be called when the access token expires.
+   */
+  onExpiredToken?: () => void
+  /**
+   * If true, the access token failed to be fetched after the maximum retries
+   */
+  failedRetry?: boolean
+}
+
+const ACCESS_TOKEN_REFRESH_INTERVAL = 3300000 // 55 minutes
+const ACCESS_TOKEN_MAX_RETRIES = 3
+
+export const AccessTokenContext = createContext<AccessTokenContextValue | undefined>({})
+
+export interface AccessTokenProviderProps extends PropsWithChildren {
+  /**
+  * Function that the provider will use to fetch the access token.
+  * @returns {Promise<string>} A promise that resolves to the access token
+  */
+  fetchToken?: () => Promise<string>
+  /**
+  * If passed, the provider will ignore the `fetchToken` function and pass this access token to all the children components.
+  */
+  accessToken?: string
+  /**
+  * Function that will be called when the access token expires.
+  */
+  onAccessTokenExpired?: () => void
+}
+
+export const AccessTokenProvider = (props: AccessTokenProviderProps) => {
+  const { children, accessToken: accessTokenFromProps, fetchToken, onAccessTokenExpired } = props
+
+  const [isLoading, setIsLoading] = useState(accessTokenFromProps == null)
+  const [fetchedToken, setFetchedToken] = useState<string | undefined>(undefined)
+  const [failedRetry, setFailedRetry] = useState(false)
+
+  const log = useLog()
+
+  const shouldFetchToken = accessTokenFromProps == null
+
+  const interval = useRef<NodeJS.Timeout>()
+
+  const fetch = useCallback(async () => {
```

Comment:
I don't think we need to use a useCallBack hook here, because we don't providing any dependincy.
and also not need to return token string. You already setting up fetchedToken with "setFetchedToken". When you need the token you can read it.